debian/key.3
.TH "Key Class Methods" 3 "18 Oct 2004" "Elektra Project" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Key Class Methods \- The Key Class and its methods. 
More...
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum \fBKeyType\fP { \fBKEY_TYPE_UNDEFINED\fP = 0, \fBKEY_TYPE_DIR\fP = 1, \fBKEY_TYPE_LINK\fP = 2, \fBKEY_TYPE_BINARY\fP = 20, \fBKEY_TYPE_STRING\fP = 40 }"
.br
.ti -1c
.RI "enum \fBKeyNamespace\fP { \fBKEY_NS_SYSTEM\fP = 1, \fBKEY_NS_USER\fP = 2 }"
.br
.ti -1c
.RI "enum \fBKeyFlags\fP { , \fBKEY_FLAG_HASTYPE\fP = 1, \fBKEY_FLAG_HASKEY\fP = 1<<1, \fBKEY_FLAG_HASDATA\fP = 1<<2, \fBKEY_FLAG_HASDOMAIN\fP = 1<<5, \fBKEY_FLAG_HASCOMMENT\fP = 1<<6, \fBKEY_FLAG_HASUID\fP = 1<<7, \fBKEY_FLAG_HASGID\fP = 1<<8, \fBKEY_FLAG_HASPRM\fP = 1<<10, \fBKEY_FLAG_HASTIME\fP = 1<<11, \fBKEY_FLAG_NEEDSYNC\fP = 1<<12, \fBKEY_FLAG_FLAG\fP = 1<<31 }"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "int \fBkeyInit\fP (Key *key)"
.br
.ti -1c
.RI "int \fBkeyClose\fP (Key *key)"
.br
.ti -1c
.RI "int \fBkeyIsInitialized\fP (const Key *key)"
.br
.ti -1c
.RI "int \fBkeyNeedSync\fP (const Key *key)"
.br
.ti -1c
.RI "u_int8_t \fBkeyGetType\fP (const Key *key)"
.br
.ti -1c
.RI "u_int8_t \fBkeySetType\fP (Key *key, u_int8_t newType)"
.br
.ti -1c
.RI "size_t \fBkeyGetDataSize\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyGetNameSize\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyGetFullNameSize\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyGetFullName\fP (const Key *key, char *returnedName, size_t maxSize)"
.br
.ti -1c
.RI "size_t \fBkeyGetName\fP (const Key *key, char *returnedName, size_t maxSize)"
.br
.ti -1c
.RI "size_t \fBkeySetName\fP (Key *key, const char *newName)"
.br
.ti -1c
.RI "size_t \fBkeyGetOwner\fP (const Key *key, char *returned, size_t maxSize)"
.br
.ti -1c
.RI "size_t \fBkeyGetOwnerSize\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeySetOwner\fP (Key *key, const char *userDomain)"
.br
.ti -1c
.RI "size_t \fBkeyGetComment\fP (const Key *key, char *returnedDesc, size_t maxSize)"
.br
.ti -1c
.RI "size_t \fBkeySetComment\fP (Key *key, const char *newComment)"
.br
.ti -1c
.RI "size_t \fBkeyGetCommentSize\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyGetString\fP (const Key *key, char *returnedString, size_t maxSize)"
.br
.ti -1c
.RI "size_t \fBkeySetString\fP (Key *key, const char *newString)"
.br
.ti -1c
.RI "size_t \fBkeyGetBinary\fP (const Key *key, void *returnedBinary, size_t maxSize)"
.br
.ti -1c
.RI "size_t \fBkeySetBinary\fP (Key *key, const void *newBinary, size_t dataSize)"
.br
.ti -1c
.RI "size_t \fBkeySetRaw\fP (Key *key, const void *newBinary, size_t dataSize)"
.br
.ti -1c
.RI "size_t \fBkeySetLink\fP (Key *key, const char *target)"
.br
.ti -1c
.RI "int \fBkeyDup\fP (const Key *source, Key *dest)"
.br
.ti -1c
.RI "uid_t \fBkeyGetUID\fP (const Key *key)"
.br
.ti -1c
.RI "int \fBkeySetUID\fP (Key *key, uid_t uid)"
.br
.ti -1c
.RI "gid_t \fBkeyGetGID\fP (const Key *key)"
.br
.ti -1c
.RI "int \fBkeySetGID\fP (Key *key, gid_t gid)"
.br
.ti -1c
.RI "mode_t \fBkeyGetAccess\fP (const Key *key)"
.br
.ti -1c
.RI "int \fBkeySetAccess\fP (Key *key, mode_t mode)"
.br
.ti -1c
.RI "time_t \fBkeyGetMTime\fP (const Key *key)"
.br
.ti -1c
.RI "time_t \fBkeyGetATime\fP (const Key *key)"
.br
.ti -1c
.RI "time_t \fBkeyGetCTime\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyGetParentSize\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyGetParent\fP (const Key *key, char *returnedParent, size_t maxSize)"
.br
.ti -1c
.RI "u_int32_t \fBkeyCompare\fP (const Key *key1, const Key *key2)"
.br
.ti -1c
.RI "size_t \fBkeyToStream\fP (const Key *key, FILE *stream, unsigned long options)"
.br
.ti -1c
.RI "int \fBkeyNameIsSystem\fP (const char *keyName)"
.br
.ti -1c
.RI "int \fBkeyNameIsUser\fP (const char *keyName)"
.br
.ti -1c
.RI "int \fBkeyIsSystem\fP (const Key *key)"
.br
.ti -1c
.RI "int \fBkeyIsUser\fP (const Key *key)"
.br
.ti -1c
.RI "int \fBkeyNameGetNamespace\fP (const char *keyName)"
.br
.ti -1c
.RI "int \fBkeyGetNamespace\fP (const Key *key)"
.br
.ti -1c
.RI "int \fBkeyIsDir\fP (const Key *key)"
.br
.ti -1c
.RI "int \fBkeyIsLink\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyNameGetRootNameSize\fP (const char *keyName)"
.br
.ti -1c
.RI "size_t \fBkeyGetRootNameSize\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyGetFullRootNameSize\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyNameGetBaseNameSize\fP (const char *keyName)"
.br
.ti -1c
.RI "size_t \fBkeyGetBaseNameSize\fP (const Key *key)"
.br
.ti -1c
.RI "size_t \fBkeyGetRootName\fP (const Key *key, char *returned, size_t maxSize)"
.br
.ti -1c
.RI "size_t \fBkeyGetFullRootName\fP (const Key *key, char *returned, size_t maxSize)"
.br
.ti -1c
.RI "size_t \fBkeyGetBaseName\fP (const Key *key, char *returned, size_t maxSize)"
.br
.ti -1c
.RI "int \fBkeySetFlag\fP (Key *key)"
.br
.ti -1c
.RI "int \fBkeyClearFlag\fP (Key *key)"
.br
.ti -1c
.RI "int \fBkeyGetFlag\fP (const Key *key)"
.br
.in -1c
.SH "DETAILED DESCRIPTION"
.PP 
The Key Class and its methods.
.PP
.PP
 To use it: 
.PP
.nf
#include <kdb.h>
.fi
.PP
A Key is the essential class that contains all key data and metadata. Key properties are:
.TP
Key name
.TP
User domain
.TP
Key value or data
.TP
Data type
.TP
Comment about the key
.TP
UID, GID and access filesystem-like permissions
.TP
Access, change and modification times
.TP
A general flag
.PP
Rules for Key Names
.PP
When using Elektra to store your application's configuration and state, please keep in mind the following rules:
.TP
You are not allowed to create keys right under \fCsystem\fP or \fCuser\fP.
.TP
You are not allowed to create folder keys right under \fCsystem\fP or \fCuser\fP. They are reserved for very essential OS subsystems.
.TP
The keys for your application, called say \fIMyApp\fP, should be created under \fCsystem/sw/MyApp\fP and/or \fCuser/sw/MyApp\fP.
.TP
It is suggested to make your application look for default keys under \fCsystem/sw/MyApp/current\fP and/or \fCuser/sw/MyApp/current\fP. This way, from a sysadmin perspective, it will be possible to copy the \fCsystem/sw/MyApp/current\fP tree to something like \fCsystem/sw/MyApp/old\fP, and keep system clean and organized.
.PP
Described here the methods to get and set, and make various manipulations in the objects of class Key. 
.SH "ENUMERATION TYPE DOCUMENTATION"
.PP 
.SS "enum KeyFlags"
.PP
Flags that can be used with keys. 
.PP
\fBSee also: \fP
.in +1c
\fBkeyCompare\fP() , \fBkdbMonitorKey\fP() the diffMask parameter , \fBkdbMonitorKeys\fP() the diffMask parameter , \fBkeyGetFlag\fP() 
.PP
\fBEnumeration values:\fP
.in +1c
.TP
\fB\fI\fIKEY_FLAG_HASTYPE\fP \fP\fP
Flag for the key type 
.TP
\fB\fI\fIKEY_FLAG_HASKEY\fP \fP\fP
Flag for the key name 
.TP
\fB\fI\fIKEY_FLAG_HASDATA\fP \fP\fP
Flag for the key data 
.TP
\fB\fI\fIKEY_FLAG_HASDOMAIN\fP \fP\fP
Flag for the key user domain 
.TP
\fB\fI\fIKEY_FLAG_HASCOMMENT\fP \fP\fP
Flag for the key comment 
.TP
\fB\fI\fIKEY_FLAG_HASUID\fP \fP\fP
Flag for the key UID 
.TP
\fB\fI\fIKEY_FLAG_HASGID\fP \fP\fP
Flag for the key GID 
.TP
\fB\fI\fIKEY_FLAG_HASPRM\fP \fP\fP
Flag for the key permissions 
.TP
\fB\fI\fIKEY_FLAG_HASTIME\fP \fP\fP
Flag for the key change time 
.TP
\fB\fI\fIKEY_FLAG_NEEDSYNC\fP \fP\fP
Flags that key needs syncronization 
.TP
\fB\fI\fIKEY_FLAG_FLAG\fP \fP\fP
General purpose flag that has semantics only to your app 
.PP
Definition at line 134 of file kdb.h.
.SS "enum KeyNamespace"
.PP
Elektra currently supported Key namespaces. 
.PP
\fBSee also: \fP
.in +1c
\fBkdbGetRootKeys\fP() , \fBkeyGetNamespace\fP() , \fBkeyNameGetNamespace\fP() 
.PP
\fBEnumeration values:\fP
.in +1c
.TP
\fB\fI\fIKEY_NS_SYSTEM\fP \fP\fP
The \fCsystem\fP keys 
.TP
\fB\fI\fIKEY_NS_USER\fP \fP\fP
The \fCuser\fP keys 
.PP
Definition at line 97 of file kdb.h.
.SS "enum KeyType"
.PP
Key data types.
.PP
Key type values grow from the semantically poor to the semantically rich. The gaps between them is for user-defined types.
.PP
If your application needs value types with more semantics, like \fCColor\fP, \fCFont\fP, etc, you can still use it. You'll have to define a new type number in the scope of your application, and force the type with \fBkeySetType\fP().
.PP
The type number is a value between 0 and 255. If you define it bigger than \fCKEY_TYPE_STRING\fP, it will be still treated as a string (in the terms of Unicode handling). If you define it between \fCKEY_TYPE_BINARY\fP and \fCKEY_TYPE_STRING\fP, Elektra will handle it as a binary, will not make Unicode handling and will save it hex-encoded.
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetType\fP() , \fBkeySetType\fP() for an example of how to define custom types 
.PP
\fBEnumeration values:\fP
.in +1c
.TP
\fB\fI\fIKEY_TYPE_UNDEFINED\fP \fP\fP
Undefined key type 
.TP
\fB\fI\fIKEY_TYPE_DIR\fP \fP\fP
A directory key 
.TP
\fB\fI\fIKEY_TYPE_LINK\fP \fP\fP
A symbolink link key. This gap is for special key meta types, that can't go into regular files. 
.TP
\fB\fI\fIKEY_TYPE_BINARY\fP \fP\fP
A binary key. This gap is for binary data types that have some semantics that somebody can invent in the future 
.TP
\fB\fI\fIKEY_TYPE_STRING\fP \fP\fP
A string key 
.PP
Definition at line 72 of file kdb.h.
.SH "FUNCTION DOCUMENTATION"
.PP 
.SS "int keyClearFlag (Key * key)"
.PP
Clear the general flag in the Key
.PP
The flag has no semantics to the library, only to your application.
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetFlag\fP() , \fBkeySetFlag\fP() 
.PP
\fBReturns: \fP
.in +1c
0, unless key is invalid. 
.PP
Definition at line 2165 of file key.c.
.SS "int keyClose (Key * key)"
.PP
Finishes the usage of a Key object.
.PP
Frees all internally allocated memory, and leave the Key object ready to be destroyed, or explicitly by a \fIfree()\fP, or a local variable dealocation. 
.PP
\fBSee also: \fP
.in +1c
\fBkeyInit\fP() for usage example 
.PP
Definition at line 149 of file key.c.
.SS "u_int32_t keyCompare (const Key * key1, const Key * key2)"
.PP
Compare 2 keys.
.PP
The returned flag array has 1s (different) or 0s (same) for each key meta info compared, that can be logically ORed with \fCKEY_FLAG_*\fP flags.
.PP
\fBReturns: \fP
.in +1c
a bit array poiting the differences 
.PP
\fBSee also: \fP
.in +1c
\fBksCompare\fP() for examples and more detailed description , \fBKeyFlags\fP 
.PP
Definition at line 1372 of file key.c.
.SS "int keyDup (const Key * source, Key * dest)"
.PP
Clone a key.
.PP
All private information of the source key will be copied, and nothing will be shared between both keys. \fBkeyClose\fP() will be used on destination key before the operation. Internal buffers will be automatically allocated on destination.
.PP
Parameters: \fP
.in +1c
.TP
\fB\fIsource\fP\fP
the source key 
.TP
\fB\fIdest\fP\fP
the new copy of the key 
.PP
\fBReturns: \fP
.in +1c
0 on success 
.PP
\fBSee also: \fP
.in +1c
\fBkeyClose\fP() , \fBkeyInit\fP() 
.PP
Definition at line 1103 of file key.c.
.SS "mode_t keyGetAccess (const Key * key)"
.PP
Return the key filesystem-like access permissions. 
.PP
\fBSee also: \fP
.in +1c
\fBkeySetAccess\fP() 
.PP
Definition at line 1221 of file key.c.
.SS "time_t keyGetATime (const Key * key)"
.PP
Get last time the key data was read from disk. 
.PP
Definition at line 1273 of file key.c.
.SS "size_t keyGetBaseName (const Key * key, char * returned, size_t maxSize)"
.PP
Calculate the basename of a key name and put it in \fCreturned\fP.
.PP
Some examples:
.TP
basename of \fCsystem/some/keyname\fP is \fCkeyname\fP
.TP
basename of \fC'user/tmp/some key'\fP is \fC'some key'\fP
.PP
Parameters: \fP
.in +1c
.TP
\fB\fIkey\fP\fP
the key to extract basename from 
.TP
\fB\fIreturned\fP\fP
a pre-allocated buffer to store the basename 
.TP
\fB\fImaxSize\fP\fP
size of the \fCreturned\fP buffer 
.PP
\fBReturns: \fP
.in +1c
number of bytes copied to \fCreturned\fP, or 0 and \fCerrno\fP is set 
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetBaseNameSize\fP() 
.PP
Definition at line 2102 of file key.c.
.SS "size_t keyGetBaseNameSize (const Key * key)"
.PP
Calculates number of bytes needed to store basename of a key.
.PP
Basenames are denoted as:
.TP
\fCsystem/some/thing/basename\fP
.TP
\fCuser:domain/some/thing/basename\fP
.PP
\fBReturns: \fP
.in +1c
number of bytes needed without ending NULL 
.PP
\fBSee also: \fP
.in +1c
\fBkeyNameGetBaseNameSize\fP() 
.PP
Definition at line 1975 of file key.c.
.SS "size_t keyGetBinary (const Key * key, void * returnedBinary, size_t maxSize)"
.PP
Get the value of a binary or string key.
.PP
Parameters: \fP
.in +1c
.TP
\fB\fIreturnedBinary\fP\fP
pre-allocated memory to store a copy of the key value 
.TP
\fB\fImaxSize\fP\fP
number of bytes of pre-allocated memory 
.PP
\fBReturns: \fP
.in +1c
the number of bytes actually copied 
.PP
\fBSee also: \fP
.in +1c
\fBkeySetBinary\fP() , \fBkeyGetString\fP() 
.PP
Definition at line 953 of file key.c.
.SS "size_t keyGetComment (const Key * key, char * returnedDesc, size_t maxSize)"
.PP
Get the key comment.
.PP
A Key comment is pretty much as a comment in a text configuration file.
.PP
Parameters: \fP
.in +1c
.TP
\fB\fIreturnedDesc\fP\fP
pre-allocated memory to copy the comments to 
.TP
\fB\fImaxSize\fP\fP
number of bytes that will fit returnedDesc 
.PP
\fBReturns: \fP
.in +1c
number of bytes written 
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetCommentSize\fP() , \fBkeySetComment\fP() 
.PP
Definition at line 702 of file key.c.
.SS "size_t keyGetCommentSize (const Key * key)"
.PP
Calculates number of bytes needed to store a key comment, including final NULL.
.PP
Use this method to allocate memory to retrieve a key comment.
.PP
\fBReturns: \fP
.in +1c
number of bytes needed 
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetComment\fP() , \fBkeySetComment\fP() 
.PP
Definition at line 775 of file key.c.
.SS "time_t keyGetCTime (const Key * key)"
.PP
Get last time the key was stated from disk. 
.PP
Definition at line 1287 of file key.c.
.SS "size_t keyGetDataSize (const Key * key)"
.PP
Returns the number of bytes of the key value
.PP
This method is used with malloc() before a \fBkeyGetString\fP() or \fBkeyGetBinary\fP().
.PP
\fBReturns: \fP
.in +1c
the number of bytes needed to store the key value 
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetString\fP() , \fBkeyGetBinary\fP() 
.PP
Definition at line 314 of file key.c.
.SS "int keyGetFlag (const Key * key)"
.PP
Get the flag from the Key.
.PP
The flag has no semantics to the library, only to your application.
.PP
\fBSee also: \fP
.in +1c
\fBkeySetFlag\fP() , \fBkeyClearFlag\fP() 
.PP
\fBReturns: \fP
.in +1c
1 if flag is set, 0 otherwise 
.PP
Definition at line 2190 of file key.c.
.SS "size_t keyGetFullName (const Key * key, char * returnedName, size_t maxSize)"
.PP
Get key full name, including the user domain name.
.PP
\fBReturns: \fP
.in +1c
number of bytes written 
.PP
Parameters: \fP
.in +1c
.TP
\fB\fIkey\fP\fP
the key object 
.TP
\fB\fIreturnedName\fP\fP
pre-allocated memory to write the key name 
.TP
\fB\fImaxSize\fP\fP
maximum number of bytes that will fit in returnedName, including the final NULL 
.PP
Definition at line 391 of file key.c.
.SS "size_t keyGetFullNameSize (const Key * key)"
.PP
Space needed to store the key name including user domain
.PP
\fBReturns: \fP
.in +1c
number of bytes needed to store key name including user domain 
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetFullName\fP() , \fBkeyGetNameSize\fP() 
.PP
Definition at line 363 of file key.c.
.SS "size_t keyGetFullRootName (const Key * key, char * returned, size_t maxSize)"
.PP
Copy to \fCreturned\fP the full root name of the key.
.PP
Some examples:
.TP
root of \fCsystem/some/key\fP is \fCsystem\fP
.TP
root of \fCuser:denise/some/key\fP is \fCuser:denise\fP
.TP
root of \fCuser/env/env1\fP is \fCuser:$USER\fP
.PP
This method is more robust then \fBkeyGetRootName\fP()
.PP
Parameters: \fP
.in +1c
.TP
\fB\fIkey\fP\fP
the key to extract root from 
.TP
\fB\fIreturned\fP\fP
a pre-allocated buffer to store the rootname 
.TP
\fB\fImaxSize\fP\fP
size of the \fCreturned\fP buffer 
.PP
\fBReturns: \fP
.in +1c
number of bytes written to \fCreturned\fP without ending NULL 
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetFullRootNameSize\fP() , \fBkeyGetRootName\fP() 
.PP
Definition at line 2048 of file key.c.
.SS "size_t keyGetFullRootNameSize (const Key * key)"
.PP
Calculates number of bytes needed to store full root name of a key.
.PP
Possible root key names are \fCsystem\fP, \fCuser\fP or \fCuser:someuser\fP. In contrast to \fBkeyGetRootNameSize\fP(), this method considers the user domain part, and you should prefer this one.
.PP
\fBReturns: \fP
.in +1c
number of bytes needed without ending NULL 
.PP
\fBSee also: \fP
.in +1c
\fBkeyNameGetRootNameSize\fP() , \fBkeyGetRootNameSize\fP() 
.PP
Definition at line 1906 of file key.c.
.SS "gid_t keyGetGID (const Key * key)"
.PP
Get the system's group ID of a key
.PP
\fBReturns: \fP
.in +1c
the system's GID of the key 
.PP
\fBSee also: \fP
.in +1c
\fBkeySetGID\fP() , \fBkeyGetUID\fP() 
.PP
Definition at line 1185 of file key.c.
.SS "time_t keyGetMTime (const Key * key)"
.PP
Get last modification time of the key on disk. 
.PP
Definition at line 1258 of file key.c.
.SS "size_t keyGetName (const Key * key, char * returnedName, size_t maxSize)"
.PP
Get abreviated key name (without user domain name)
.PP
\fBReturns: \fP
.in +1c
number of bytes written 
.PP
Parameters: \fP
.in +1c
.TP
\fB\fIkey\fP\fP
the key object 
.TP
\fB\fIreturnedName\fP\fP
pre-allocated memory to write the key name 
.TP
\fB\fImaxSize\fP\fP
maximum number of bytes that will fit in returnedName, including the final NULL 
.PP
Definition at line 430 of file key.c.
.SS "size_t keyGetNameSize (const Key * key)"
.PP
Space needed to store the key name without user domain
.PP
\fBReturns: \fP
.in +1c
number of bytes needed to store key name without user domain 
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetName\fP() , \fBkeyGetFullNameSize\fP() 
.PP
Definition at line 343 of file key.c.
.SS "int keyGetNamespace (const Key * key)"
.PP
Return the namespace of a key
.PP
Currently valid namespaces are KEY_NS_SYSTEM and KEY_NS_USER.
.PP
\fBReturns: \fP
.in +1c
KEY_NS_SYSTEM, KEY_NS_USER or 0 
.PP
\fBSee also: \fP
.in +1c
keyNameGetNameSpace() , \fBkeyIsUser\fP() , \fBkeyIsSystem\fP() , \fBKeyNamespace\fP 
.PP
Definition at line 1796 of file key.c.
.SS "size_t keyGetOwner (const Key * key, char * returned, size_t maxSize)"
.PP
Return the user domain of the key.
.TP
Given \fCuser:someuser/\fP..... return \fCsomeuser\fP
.TP
Given \fCuser:some.user/\fP.... return \fCsome.user\fP
.TP
Given \fCuser/\fP.... return the current user
.PP
Only \fCuser/\fP... keys have user domains. For \fCsystem/\fP... keys (that doesn't have user domains) nothing is returned.
.PP
Although usually the same, the user domain of a key is not related to its UID. User domains are related to WHERE the key is stored on disk, while UIDs are related to access controls of a key.
.PP
Parameters: \fP
.in +1c
.TP
\fB\fIkey\fP\fP
the object to work with 
.TP
\fB\fIreturned\fP\fP
a pre-allocated space to store the owner 
.TP
\fB\fImaxSize\fP\fP
maximum number of bytes that fit returned 
.PP
\fBReturns: \fP
.in +1c
number of bytes written to buffer 
.PP
\fBSee also: \fP
.in +1c
\fBkeySetName\fP() , \fBkeySetOwner\fP() , \fBkeyGetFullName\fP() 
.PP
Definition at line 606 of file key.c.
.SS "size_t keyGetOwnerSize (const Key * key)"
.PP
Return the size of the user domain of the Key.
.PP
\fBReturns: \fP
.in +1c
number of bytes 
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetOwner\fP() 
.PP
Definition at line 632 of file key.c.
.SS "size_t keyGetParent (const Key * key, char * returnedParent, size_t maxSize)"
.PP
Copy this key's parent name into a pre-allocated buffer.
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetParentSize\fP() 
.PP
Parameters: \fP
.in +1c
.TP
\fB\fIreturnedParent\fP\fP
pre-allocated buffer to copy parent name to 
.TP
\fB\fImaxSize\fP\fP
number of bytes pre-allocated 
.PP
Definition at line 1347 of file key.c.
.SS "size_t keyGetParentSize (const Key * key)"
.PP
Get the number of bytes needed to store this key's parent name. 
.PP
\fBSee also: \fP
.in +1c
\fBkeyGetParent\fP() 
.PP
Definition at line 1303 of file key.c.
.SS "size_t keyGetRootName (const Key * key, char * returned, size_t maxSize)"
.PP
Copy to \fCreturned\fP the root name of the key.
.PP
Some examples:
.TP
root of \fCsystem/some/key\fP is \fCsystem\fP
.TP

NEWS
0.7.0

Finally it is done, the 0.7.0 API is finished.
The API covers everything shown in kdb.h and all
releases in 0.7.0 will be 100% ABI and API compatible
supported for at least a year.

Some bugs might break that rule, but there should not be
left many anymore, because every method is at least checked
twice and most have hundred of test cases.
Please report any bugs you find to http://bugs.libelektra.org
- portability and build bugs
- mistakes, misspelling in documentation

This is not valid for kdb-tools, there are lots of bugs
remaining there and a rewrite would be very useful.

For that reason there will be a separate release for
kdb-tools, libelektratools and the backend interface.

The stable APIs of libelektratools and the backend interface
will be announced separately, but they will be part of 0.7.x.

kdb-tool and kdbedit and so will also have separate
releases, but because they are not a library there
is no API or ABI issue there.

Pages related to 0.7.0:
http://www.libelektra.org/GetStartedMounting
http://www.libelektra.org/Tutorial
http://www.libelektra.org/API-Changes0.7
http://www.libelektra.org/Changes0.7

Marketing issues:
- Freshmeat announcment
- Sourceforge upload
- API docu upload
- forums
- newsgroups
- other mailinglists

Download it from:
http://www.markus-raab.org/ftp/elektra-0.7.0.tar.gz
ftp://www.markus-raab.org/elektra-0.7.0.tar.gz







---------------




0.7.0rc5 The last release candidate before 0.7.0

I am proud to present the finished  0.7.0 external API in
kdb.h [0].

A lot changed to 0.6.10 but the fundamentals are clearer then
ever before: 3 Classes with their methods, a clear, easy and
mature API. It is very hard to use it wrong but you get
best results with very short code:

http://www.libelektra.org/Tutorial

Other parts are not finished, like libelektratools, the
kdb-tool and the backend interface. They are not part of
the official 0.7 release and get their own release later.

The reasons to release within the next week are:
 - The API is finished
 - Oyranos needs a stable release from elektra
 - the last release is long ago
 - unstable has been much more stable than stable since some time
 - there have been no commits to the stable branch since
   unstable existed
 - Elektra grows just too complex to wait for all parts to finish

Now some highlights from one month of work from 333 ChangeLog
lines:
 - many many bugs were fixed
 - the core completely compiles with -ansi -pedantic -std=c99
 - all code compiles with -Wall -Werror
 - the test-suite is now complete
 - much cleaner infrastructure within the code
 - many autoconf/automake issues fixed
 - static, dynamic building, linking with defaults backends
   everything works!
 - many new options to stat and remove all keys and so on..
 - environment variables and passwd database allow subtle
   influence of where the user configuration resides
 - kdb info now prints where configuration lies on hard disc




---------------




0.7.0rc4 - 12.05.2008
I am proud to present the fourth release candidate of elektra
0.7.0.

It is now possible to enable and disable Backends and Bindings
with configure switches. The documentation was improved.
The c++ binding now basically works, see the tests for examples.
Ini and Berkleydb backends now compile again.
There were a lot of bug fixes, like mntent.h fix for Mac OS X.




---------------




0.7.0rc3 - 24.04.2008
Call for 0.7 finish

Unfortunately there is not much response about the 0.7 release
candidates but there is still a lot to do to have a mature
library worth to be announced as stable.

Specifically there are following things to do and we really need
help to achieve that faster than end of this year:

There are now only 3 backends working well tested but undistinguished
in handling error scenarios. At least 3 more Backends are needed
also tested in real world applications. Favoured are existing but
not ported backends like berkleydb, ini, winregistry and uniconf.

Working bindings have a lot meaning in how complete and useful
the API is. It would be preferable to have at least 3 bindings
working with at least an example application like kdb-tool.
They could be cpp, python and one open.

A lot of bug squashing is left open. Please don't hesitate to open
and fix more bugs on http://bugs.libelektra.org.

Documentation is not yet completely updated to 0.7. Especially the
homepage has a lot of outdated information, but also the API documentation
needs another eye on it and the tutorial needs a rewrite.

There might be some licencing issues in some parts of elektra,
everything should be BSD, if you find something please open a bug.

A partial blocker is the kdb-tool, where a rewrite would be nice but
is not necessary for 0.7 (the library counts not the environment).

Other things like portability are taken seriously but are not blockers,
they can be fixed later too.

What is really good working in elektra is the infrastructure like
svn, homepage, bug tracking, many thanks to José Luis Tallón for
that!

Now some words about 0.7.0rc3:

Version macros were added to have static information about
the elektra version compiled against. There were many
enhancements in the C++ Binding, the Key Class is now
quite well complete.

While writing the C++ binding I realized that there is need
for keyInc() the opposite of keyDel() to increase the
viability of a key object.

keyVNew() and ksVNew() were added to make it possible to pass
the va_list from the C++ binding.

The blocker bug that libelektra assumes system/elektra/mountpoints
to exist was solved, many thanks to Patrick Sabin. The tests
now work without any preloading without memory leaks.

And lots of bug fixes (Many thanks to Kai-Uwe Behrmann):
- sed changed for macosx compatibility
- fixed signature ssize_t ksToStream
- also set CXXFlags
- fix wrong parent at end in ksToStream
- fix return value of successful kdb import
- string is default type for xml without type=""
- fix extra_dist some missing files

example/key.c
#include <kdb.h>

/**key.c
 * This is comprehensive Test for the
 * whole Elektra api (Key and KeySets).
 * main() can be used to start the
 * wished test-functions.
 */

void useKeySet()
{
	KeySet *ks=ksNew();

	ksAppend(ks,keyNew(KEY_SWITCH_END));       // an empty key
	    
	ksAppend(ks,keyNew("user/sw",              // a simple key
	    KEY_SWITCH_END));                      // no more args
	    
	ksAppend(ks,keyNew("system/sw",
	    KEY_SWITCH_NEEDSYNC,                   // a key retrieved from storage
	    KEY_SWITCH_END));                      // end of args               
	    
	ksAppend(ks,keyNew("user/tmp/ex1",
	    KEY_SWITCH_VALUE,"some data",          // with a simple value
	    KEY_SWITCH_END));                      // end of args
	    
	ksAppend(ks,keyNew("user/tmp/ex2",
	    KEY_SWITCH_VALUE,"some data",          // with a simple value
	    KEY_SWITCH_MODE,0777,                  // permissions
	    KEY_SWITCH_END));                      // end of args
	    
	ksAppend(ks,keyNew("user/tmp/ex3",
	    KEY_SWITCH_TYPE,KEY_TYPE_LINK,         // only type
	    KEY_SWITCH_VALUE,"system/mtp/x",       // link destination
	    KEY_SWITCH_MODE,0654,                  // weird permissions
	    KEY_SWITCH_END));                      // end of args
	    
	/*ksAppend(ks,keyNew("user/tmp/ex4",
	    KEY_SWITCH_TYPE,KEY_TYPE_BINARY,       // key type and value size (because it is binary)
	    KEY_SWITCH_DOMAIN,"root",              // owner (not uid) is root
	    KEY_SWITCH_VALUE,"some data",          // value that will be truncated
	    KEY_SWITCH_COMMENT,"value is truncated",
	    KEY_SWITCH_UID,0,                      // root uid
	    KEY_SWITCH_END));                      // end of args*/
	    
	ksAppend(ks,keyNew("user/env/alias/ls",    // a key we know we have
	    KEY_SWITCH_NEEDSYNC,                   // retrieve from storage
	    KEY_SWITCH_END));                      // do nothing more
	    
	ksAppend(ks,keyNew("user/env/alias/ls",    // same key
	    KEY_SWITCH_NEEDSYNC,                   // retrieve from storage
	    KEY_SWITCH_DOMAIN,"root",              // set new owner (not uid) as root
	    KEY_SWITCH_COMMENT,"new comment",      // set new comment
	    KEY_SWITCH_END));                      // end of args
	    
	ksToStream(ks,stdout,KDB_O_XMLHEADERS);
	
	ksDel(ks);
}	

int main(int argc, char **argv) {
	useKeySet();
	
	return 0;
}

